README.md
# forTheTraveler

v.0.8.4

### Данное SPA является итоговым проектом IT-Academy по курсу "Веб приложение на Javascript".

## Технические нюансы веб-приложения:

- Данные о курсах валют получены от API "Национального банка Республики Беларусь" (nbrb.by).
- Запросы к API осуществляются в отдельном потоке с помощью Worker.
- Для визуализации графиков используется библиотека Highcharts.
- Стилизация некоторых элементов SPA выполнена с помощью библиотеки Bootstrap.
- Для придания некоторой нативности данному приложению, был использован Service Worker.
- Для анализа качества кода, был использован ESLint с конфигурациями eslint-config-airbnb-base.

## Навигация по веб-приложению:

- Вкладка "Главная страница" - содержится краткая информация о данном веб-приложении.
- Вкладка "Список курсов валют" - таблица с курсами валют на выбранную дату.
- Вкладка "Конвертер валют" - интерфес для конвертации между двумя выбранными валютами.
- Вкладка "Анализ валют" - анализ изменения курса валюты за выбранный промежуток времени.

Link : [Netlify](https://forthetraveler.netlify.app/)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on forthetraveler.netlify.app(https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.8.4] - 2020-12-01
### Changed

- Corrected output of average exchange rate data by months.


## [0.8.3] - 2020-12-01
### Changed

- Corrected the output of the average exchange rate data by year.


## [0.8.2] - 2020-11-30
### Changed

- Changed the function to more accurately calculate the average value of currencies.
- Fixed bug with cropped graph.


## [0.8.1] - 2020-11-29
### Changed

- Removed unused functions.


## [0.8.0] - 2020-11-29
### Changed

- Fixed bug with missing dates.
- The correct work of the serwice worker is configured.

## [0.7.9] - 2020-11-28
### Changed

- Сonverted requests to obtain data on currencies, taking into account the changing ID for currencies.
- Considered the logic of missing exchange rates for some dates.


## [0.7.8] - 2020-11-24
### Changed

- Еliminated possible errors when entering parameters that did not correspond to the requests.


## [0.7.7] - 2020-11-24
### Changed

- Еliminated possible errors when entering parameters that did not correspond to the requests.
- Work of the site as a full-fledged PWA.


## [0.7.6] - 2020-11-24
### Changed

- The list of files for loading into the cache has been expanded.
- Work of the site as a full-fledged PWA.
- Returned screen orientation view for mobile devices "landscape".


## [0.7.5] - 2020-11-23
### Changed

- Еliminated possible errors when entering parameters that did not correspond to the requests.
- Work of the site as a full-fledged PWA.


## [0.7.4] - 2020-11-23
### Changed

- Fixed errors in the manifest.json file.
- Work of the site as a full-fledged PWA.


## [0.7.3] - 2020-11-23
### Changed

- Correct manifest.


## [0.7.2] - 2020-11-23
### Changed

- Correct name.


## [0.7.1] - 2020-11-23

### Changed

- Changed serwice worker.
- Added manifest.json.
- Correct analizCur.js.


## [0.7.0] - 2020-11-18
### Added

- It became possible to view the average exchange rate for the period.
- To analyze the quality of the code, ESLint was used with the eslint-config-airbnb-base configurations.

### Changed

- Minimally changed design.
- added functionality to eliminate the difference in ID for a different period of time.


## [0.6.0] - 2020-11-03
### Added

- Added the ability to select multiple currencies for analysis, as well as the ability to select different types of periodicity.
- To analyze the quality of the code, ESLint was used with the eslint-config-airbnb-base configurations.
- Added .gitignore to filter files uploaded to github.

### Changed

- Minimally changed design.
- Fixed bugs with displaying charts (currencies were incorrectly indicated).

### Removed

- Unnecessary pictures and icons have been removed.

## [0.5.3] - 2020-10-30

### Changed

- Fixed a bug with incorrect display of charts.

## [0.5.2] - 2020-10-29

### Changed

- Correct dates in the charts.

## [0.5.1] - 2020-10-29

### Changed

- Minimally changed design.
- It became possible to select several currencies for analysis.

## [0.5.0] - 2020-10-27
### Added

- In the Currency Analysis tab, you can now select an interval.

### Changed

- The structure of the SPA menu has been changed.
- To minimize requests to the API when converting currencies, a state simulation was implemented to store the exchange rate after the first request to the API.
- Added a timeout for a spinner when loading data.

### Removed

- Scroll to content when clicking on tabs.

## [0.4.0] - 2020-10-17
### Added

- Added a spinner when loading data from API.
- Added page style.
- Added service worker.

### Changed

- Set restrictions when choosing dates in input type = date.

## [0.3.2] - 2020-10-09

- Moved "then" to a separate function in "worker.js".
- In "main.js" the tab switching function has been changed.
- In "analizCur.js" functions have been changed to avoid creating additional context.

### Changed

- Fixed a bug with drawing graphs of currency changes.

## [0.3.1] - 2020-10-07

### Changed

- Fixed a bug with drawing graphs of currency changes.

## [0.3.0] - 2020-10-07
### Added

- Added "Main page" tab;

### Changed

- Сhanged the style of the navigation menu.
- The request to the worker is formed as an object.
- Changed currency conversion table.

## [0.2.0] - 2020-09-30
### Added

- File analizCur.js, allows you to analyze the change in the exchange rate for a selected period of time.
- File charts.js, allows you to visualize graphs of currency changes.

### Changed

- Added the analiz block to index.html and included the highcharts.com library.
- Added classes to style.css for visualizing charts.
- Added the ability to get worker.js data using two urls.

## [0.1.0] - 2020-09-28
### Added

- File main.js which allows you to change the content in the SPA, depending on the selected tab.
- File worker.js, with which a request is made to the nbrb.by API to receive data.
- File curent.js, creates a table with exchange rates using data received from worker.
- File convert.js, makes it possible to convert the currencies that it receives from the worker.
- All the above files are added to the 'js' folder.
- Folder 'css' to which the style.css file with styles is added.
- Folder 'img' to which a picture with the background of the page is added.

### Changed

- Content index.html, in particular added: page menu, page blocks corresponding to menu tabs, all files from the 'js', 'css' folders are included, and the highcharts library is also connected.
- File CHANGELOG.md in accordance with the adjustments and development of the project.
- File README.md in accordance with the adjustments and development of the project.

## [0.0.2] - 2020-09-09
### Changed

- Fixed bug in link to Netlify, in file README.md.

## [0.0.1] - 2020-09-09
### Added


js/analizCur.js
/* eslint-disable no-loop-func */
/* eslint-disable no-else-return */
const startDateRates = document.querySelector('#startDateRates');
const endDateRates = document.querySelector('#endDateRates');
radioInput.forEach((el) => {
  el.addEventListener('click', () => {
    radioInput.forEach((elem) => {
      document.querySelector(`.${elem.value}`).style.display = 'none';
    });
    if (el.checked) {
      const selectCur = document.querySelector('#choiseCur');
      if (el.value === 'rates') {
        selectCur.removeAttribute('multiple');
      } else {
        selectCur.setAttribute('multiple', 'multiple');
      }
      document.querySelector(`.${el.value}`).style.display = 'block';
    }
  });
});

radioInputTime.forEach((el) => {
  el.addEventListener('click', () => {
    radioInputTime.forEach((elem) => {
      document.querySelector(`#${elem.value}`).style.display = 'none';
    });
    if (el.checked) {
      document.querySelector(`#${el.value}`).style.display = 'block';
    }
  });
});

function pickMaxDate() {
  startDate.setAttribute('max', new Date(endDate.value).toISOString().split('T')[0]);
}

function pickMinDate() {
  endDate.setAttribute('min', new Date(startDate.value).toISOString().split('T')[0]);
}

function startAnalizCur() {
  if (selectAnaliz.value === '') {
    try {
      myWorker.postMessage(JSON.stringify('get_cur_data'));
      myWorker.onmessage = (e) => {
        const data = JSON.parse(e.data);
        sessionStorage.setItem('startAnalizCur', JSON.stringify(data));
        createOptionsConvert(data, selectAnaliz);
      };
    } catch (error) {
      const sessionData = sessionStorage.getItem('startAnalizCur');
      const funData = JSON.parse(sessionData);
      createOptionsConvert(funData, selectAnaliz);
    }
  }
  startDate.setAttribute('max', today);
  endDate.setAttribute('max', today);
}

function choiseInterval() {
  const selectInterval = document.querySelector('#choise_interval').value;
  let dateNow = Date.now();
  let i = 0;
  dates = [];
  while (i < selectInterval) {
    dates.unshift(new Date(dateNow).toISOString().substr(0, 10));
    dateNow -= 24 * 60 * 60 * 1000;
    i++;
  }
  prepData();
}

function showAnaliz() {
  if (startDate.value && endDate.value) {
    rangeDate(startDate.value, endDate.value);
    prepData();
  } else {
    return alert('Введена некорректная дата');
  }
}

function changeDateInput() {
  const intervalRates = document.querySelector('#choise_intervalRates').value;
  if (intervalRates === '28') {
    startDateRates.type = 'month';
    endDateRates.type = 'month';
  } else if (intervalRates === '365') {
    const today = new Date().getFullYear();
    startDateRates.type = 'number';
    endDateRates.type = 'number';
    endDateRates.hasAttribute('required');
    endDateRates.setAttribute('max', today);
    endDateRates.setAttribute('min', '1992');
    startDateRates.setAttribute('max', today);
    startDateRates.setAttribute('min', '1992');
    startDateRates.value = '1992';
    endDateRates.value = today;
  } else {
    startDateRates.type = 'date';
    endDateRates.type = 'date';
  }
}

function pickMaxDateRate() {
  startDateRates.setAttribute('max', new Date(endDateRates.value).toISOString().split('T')[0]);
}

function pickMinDateRate() {
  endDateRates.setAttribute('min', new Date(startDateRates.value).toISOString().split('T')[0]);
}

function prepData() {
  const arrVal = [];
  const opt = document.querySelectorAll('#choiseCur option');
  opt.forEach((el) => {
    if (el.selected) arrVal.push(el.value);
  });
  if (arrVal.length === 0) return alert('Ни одна валюта не выбрана');
  const objDate = {
    current: arrVal,
    date: dates
  };
  spinnerPage.render();
  myWorker.postMessage(JSON.stringify(objDate));
  myWorker.onmessage = async (e) => {
    const data = await JSON.parse(e.data);
    const info = await correctData(data);
    await createGraf(info);
    spinnerPage.handleClear();
  };
}

function showAnalizRates() {
  const arrVal = [];
  const opt = document.querySelectorAll('#choiseCur option');
  opt.forEach((el) => {
    if (el.selected) arrVal.push(el.value);
  });
  const intervalRates = document.querySelector('#choise_intervalRates').value;
  const dateStartRates = document.querySelector('#startDateRates').value;
  const dateEndRates = document.querySelector('#endDateRates').value;
  if (dateStartRates && dateEndRates && intervalRates) {
    if (dateEndRates.length === 4) {
      rangeDate(`${dateStartRates}-01-01`, `${dateEndRates}-12-31`);
    } else if (intervalRates === '28') {
      rangeDate(`${dateStartRates}-01`, `${dateEndRates}-31`);
    } else {
      rangeDate(dateStartRates, dateEndRates);
    }
    const objDate = {
      current: arrVal,
      date: dates
    };
    spinnerPage.render();
    myWorker.postMessage(JSON.stringify(objDate));
    myWorker.onmessage = async (e) => {
      const data = await JSON.parse(e.data);
      const info = await correctData(data);
      const rates = await prepRates(info, intervalRates, dateStartRates, dateEndRates);
      await createGraf(rates);
      await spinnerPage.handleClear();
    };
  } else {
    return alert('Введена некорректная дата');
  }
}

async function prepRates(data, interval, start, end) {
  const arrDates = [];
  const arrYears = [];
  let newDate = [];
  const res = data[1].map((el) => {
    const rates = [];
    if (interval !== '365' && interval !== '28') {
      while (data[0].length > interval) {
        arrDates.push(data[0][interval - 1]);
        const newArr = el.data.slice(0, interval);
        const result = newArr.reduce((sum, current) => (sum + current), 0);
        rates.push(Number((result / interval).toFixed(3)));
        data[0].splice(0, interval);
        el.data.splice(0, interval);
      }
      return { name: el.name, data: rates };
    } else if (interval === '28') {
      while (data[0].length >= interval) {
        const year = dates[0].slice(0, 4);
        const mounth = dates[0].slice(6, 7);
        const newInterval = new Date(year, mounth, 0).getDate();
        if (el.data.length < newInterval * 2 - 1) {
          arrDates.push(data[0][newInterval - 2]);
          const newArr = el.data.slice(0, newInterval);
          const result = newArr.reduce((sum, current) => (sum + current), 0);
          rates.push(Number((result / newInterval).toFixed(3)));
          data[0].splice(0, newInterval);
          el.data.splice(0, newInterval);
        } else {
          arrDates.push(data[0][newInterval - 1]);
          const newArr = el.data.slice(0, newInterval);
          const result = newArr.reduce((sum, current) => (sum + current), 0);
          rates.push(Number((result / newInterval).toFixed(3)));
          data[0].splice(0, newInterval);
          el.data.splice(0, newInterval);
        }
      }
      newDate = arrDates.map((elem) => {
        const day = new Date(elem);
        const optionDate = { year: 'numeric', month: 'long' };
        return day.toLocaleString('ru-RU', optionDate);
      });
      return { name: el.name, data: rates };
      // eslint-disable-next-line no-else-return
    } else {
      let i = Number(start);
      while (i <= Number(end)) {
        arrYears.push(i);
        i += 1;
      }
      let j = 0;
      while (j < arrYears.length) {
        const sdate = new Date(`${arrYears[j]}-01-01`);
        const edate = new Date(`${arrYears[j]}-12-31`);
        // eslint-disable-next-line max-len
        const newInterval = Math.round((Date.parse(edate) - Date.parse(sdate)) / (1000 * 60 * 60 * 24)) + 1;
        if (newInterval > el.data.length) {
          const newArr = el.data.slice(0, el.data.length);
          const result = newArr.reduce((sum, current) => (sum + current), 0);
          rates.push(Number((result / el.data.length).toFixed(3)));
          el.data.splice(0, el.data.length);
          j++;
        } else {
          const newArr = el.data.slice(0, newInterval);
          const result = newArr.reduce((sum, current) => (sum + current), 0);
          rates.push(Number((result / newInterval).toFixed(3)));
          el.data.splice(0, newInterval);
          j++;
        }
      }
      return { name: el.name, data: rates };
    }
  });
  switch (interval) {
    case '1':
      return [arrDates, res];
    case '7':
      return [arrDates, res];
    case '28':
      return [newDate, res];
    case '365':
      return [arrYears, res];
  }
}

function rangeDate(sDate, eDate) {
  const dStart = new Date(sDate);
  const dEnd = new Date(eDate);
  const daysLag = Math.ceil(Math.abs(dEnd.getTime() - dStart.getTime()) / (1000 * 3600 * 24));
  let dateParse = Date.parse(eDate);
  let i = 0;
  dates = [];
  while (i < daysLag + 1) {
    dates.unshift(new Date(dateParse).toISOString().substr(0, 10));
    dateParse -= 24 * 60 * 60 * 1000;
    i++;
  }
}

async function correctData(data) {
  const resultCur = [];
  for (const key in data) {
    const arrReturn = [];
    data[key].dateArr.forEach((el, index) => {
      const idx = dates.indexOf(el);
      arrReturn[idx] = data[key].resultArr[index];
    });
    for (let i = 0; i < arrReturn.length; i++) {
      if (!arrReturn[i]) arrReturn[i] = 0;
    }
    resultCur.push({
      turboThreshold: dates.length,
      name: key,
      data: arrReturn
    });
  }
  return [dates, resultCur];
}
